examples/examples/z_bytes.rs

```diff
@@ -0,0 +1,143 @@
+    // let encoding = Encoding::ZENOH_BYTES;
+
+    // Writer & Reader
+    // serialization
+    let mut bytes = ZBytes::empty();
+    let mut writer = bytes.writer();
+    let i1 = 1234_u32;
+    let i2 = String::from("test");
+    let i3 = vec![1, 2, 3, 4];
+    writer.serialize(i1);
+    writer.serialize(&i2);
+    writer.serialize(&i3);
+    // deserialization
+    let mut reader = bytes.reader();
+    let o1: u32 = reader.deserialize().unwrap();
+    let o2: String = reader.deserialize().unwrap();
+    let o3: Vec<u8> = reader.deserialize().unwrap();
+    assert_eq!(i1, o1);
+    assert_eq!(i2, o2);
+    assert_eq!(i3, o3);
+
+    // Tuple
+    let input = (1234_u32, String::from("test"));
+    let payload = ZBytes::serialize(input.clone());
+    let output: (u32, String) = payload.deserialize().unwrap();
+    assert_eq!(input, output);
+
+    // Iterator
+    let input: [i32; 4] = [1, 2, 3, 4];
+    let payload = ZBytes::from_iter(input.iter());
+    for (idx, value) in payload.iter::<i32>().enumerate() {
+        assert_eq!(input[idx], value.unwrap());
+    }
+
+    // HashMap
+    let mut input: HashMap<usize, String> = HashMap::new();
+    input.insert(0, String::from("abc"));
+    input.insert(1, String::from("def"));
+    let payload = ZBytes::from(input.clone());
+    let output = payload.deserialize::<HashMap<usize, String>>().unwrap();
+    assert_eq!(input, output);
+
+    // JSON
+    let data = r#"
+    {
+        "name": "John Doe",
+        "age": 43,
+        "phones": [
+            "+44 1234567",
+            "+44 2345678"
+        ]
+    }"#;
+    let input: serde_json::Value = serde_json::from_str(data).unwrap();
+    let payload = ZBytes::try_serialize(input.clone()).unwrap();
+    let output: serde_json::Value = payload.deserialize().unwrap();
+    assert_eq!(input, output);
+    // Corresponding encoding to be used in operations like `.put()`, `.reply()`, etc.
+    // let encoding = Encoding::APPLICATION_JSON;
+
+    // YAML
+    let data = r#"
+        name: "John Doe"
+        age: 43
+        phones:
+          - "+44 1234567"
+          - "+44 2345678"
+    "#;
+    let input: serde_yaml::Value = serde_yaml::from_str(data).unwrap();
+    let payload = ZBytes::try_serialize(input.clone()).unwrap();
+    let output: serde_yaml::Value = payload.deserialize().unwrap();
+    assert_eq!(input, output);
+    // Corresponding encoding to be used in operations like `.put()`, `.reply()`, etc.
+    // let encoding = Encoding::APPLICATION_YAML;
+
+    // Protobuf
+    use prost::Message;
+    #[derive(Message, Eq, PartialEq)]
+    struct EntityInfo {
+        #[prost(uint32)]
+        id: u32,
+        #[prost(string)]
+        name: String,
+    }
+    let input = EntityInfo {
+        id: 1234,
+        name: String::from("John Doe"),
+    };
+    let payload = ZBytes::from(input.encode_to_vec());
+    let output =
+        EntityInfo::decode(Cursor::new(payload.deserialize::<Vec<u8>>().unwrap())).unwrap();
```

Comment:
Could we use a `payload.reader()`?

Reply:
In addition to that it would make sense to show in the example how to use a `ZBytesReader` and `ZBytesWriter`.

Reply:
I'm not sure whether it's better or not.

```rust
// serialize
let mut bytes = ZBytes::empty();
let mut writer = bytes.writer();
writer.serialize(input.encode_to_vec());
// deserialize
let mut reader = bytes.reader();
let output: EntityInfo =
    EntityInfo::decode(Cursor::new(reader.deserialize::<Vec<u8>>().unwrap())).unwrap();
```

BTW, we've already had `ZBytesReader` and `ZBytesWriter` here.
https://github.com/eclipse-zenoh/zenoh/pull/1180/files#diff-3d491564591b569630ebbbe501a1d6331086fc9b3c17399285f3e97cbb3f4a5dR52

Reply:
I was referring to the fact of using something like 

```rust
let output: EntityInfo =
    EntityInfo::decode(payload.reader()).unwrap();
```

To avoid potential memory allocation.

Reply:
I think we need to have `impl TryFrom<ZBytes> for bytes::Buf` to support this. I can give it a try.

Reply:
Hmmm, after having a further look, `bytes:Buf` is the trait that needs to be implemented in `ZBytes`. We can rely on `ZBufReader` to implement it, but it seems a little difficult to implement `chunk()` function since it needs to return the all remaining data. Our data might be stored in several ZSlices, so I'm not sure whether there is a way to do so.
https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html#tymethod.chunk

Reply:
Ok, then let's just deserialize it to 

```rust
let output: EntityInfo = EntityInfo::decode(Cursor::new(reader.deserialize::<Cow<[u8]>>().unwrap())).unwrap();
```